### PR TITLE
feat: handle `typeof`

### LINKED ISSUE #632: Error: Unknown type: TypeOperator
I'm having an issue when trying to define a type with keys of an enum. 

## Sorting

* **I'm submitting a ...**
  - [x] bug report
  - [ ] feature request
  - [ ] support request

* I confirm that I
  - [x] used the [search](https://github.com/lukeautry/tsoa/search?type=Issues) to make sure that a similar issue hasn't already been submit

## Expected Behavior

```ts
enum Enum {
  OK,
  KO
}
export interface SerializedDatasource {
  enumKeys: keyof typeof Enum // which produces 'OK' | 'KO'
}
```

```yaml
enumKeys:
  type: string
  enum:
    - OK
    - KO
```

## Current Behavior

```
There was a problem resolving type of 'SerializedDatasource'.
Generate swagger error.
 Error: Unknown type: TypeOperator
At: src/controllers/test.ts:11:11.
This was caused by 'enumKeys: keyof typeof Enum'
    at new GenerateMetadataError (/private/tmp/tsoa-bug/packages/api/node_modules/tsoa/dist/metadataGeneration/exceptions.js:21:28)
    at TypeResolver.resolve (/private/tmp/tsoa-bug/packages/api/node_modules/tsoa/dist/metadataGeneration/typeResolver.js:243:19)
    at TypeResolver.propertyFromSignature (/private/tmp/tsoa-bug/packages/api/node_modules/tsoa/dist/metadataGeneration/typeResolver.js:645:119)
    at /private/tmp/tsoa-bug/packages/api/node_modules/tsoa/dist/metadataGeneration/typeResolver.js:612:161
    at Array.map (<anonymous>)
```

## Context (Environment)
Version of the library: "3.0.4"
Version of NodeJS: v12.14.1
* Confirm you were using yarn not npm: [x]

### PR DESCRIPTION
**Pull request by Eywek**
### All Submissions:

* [x] Have you followed the guidelines in our [Contributing](https://github.com/lukeautry/tsoa/tree/master/docs/CONTRIBUTING.md) document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/lukeautry/tsoa/pulls) for the same update/change?
* [x] Have you written unit tests?
* [x] Have you written unit tests that cover the negative cases (i.e.: if bad data is submitted, does the library respond properly)?
* [x] This PR is associated with an existing issue?

**Closing issues**

Closes #632

### CODE REVIEW

File: src/metadataGeneration/typeResolver.ts [hunk 1]
```diff
@@ -207,6 +207,11 @@ export class TypeResolver {
       }
     }
 
+    if (ts.isTypeOperatorNode(this.typeNode)) {
```
Comment (WoH) [inline src/metadataGeneration/typeResolver.ts]:
```suggestion
    if (ts.isTypeOperatorNode(this.typeNode) && this.typeNode.operator === ts.SyntaxKind.KeyOfKeyword) {
```

File: src/metadataGeneration/typeResolver.ts [hunk 2]
```diff
@@ -207,6 +207,11 @@ export class TypeResolver {
...
+      const type = this.current.typeChecker.getTypeFromTypeNode(this.typeNode);
+      return new TypeResolver(this.current.typeChecker.typeToTypeNode(type)!, this.current, this.typeNode, this.context, this.referencer).resolve();
```
Comment (WoH) [inline src/metadataGeneration/typeResolver.ts]: this will most likely fail for more complex scenarios, need to try/catch

Comment (Eywek) [reply src/metadataGeneration/typeResolver.ts]: okay, and what do we want to do if it fails?

Comment (WoH) [reply src/metadataGeneration/typeResolver.ts]: some error message that highlights the node that causes the issue:
```ts
 throw new GenerateMetadataError(
              `Could not determine the keys of type ${this.current.typeChecker.typeToString(this.typeNode.type)} `,
              this.typeNode,
            );
```

File: src/metadataGeneration/typeResolver.ts [hunk 3]
```diff
@@ -207,6 +207,16 @@ export class TypeResolver {
       }
     }
 
+    if (ts.isTypeOperatorNode(this.typeNode) && this.typeNode.operator === ts.SyntaxKind.KeyOfKeyword) {
+      const type = this.current.typeChecker.getTypeFromTypeNode(this.typeNode);
+      const indexedTypeName = this.current.typeChecker.typeToTypeNode(type);
```
Comment (WoH) [inline src/metadataGeneration/typeResolver.ts]:
```suggestion
      const indexedTypeName = this.current.typeChecker.typeToString(this.current.typeChecker.getTypeFromTypeNode(this.typeNode.type));
```

Comment (WoH) [inline src/metadataGeneration/typeResolver.ts]:
```suggestion
      const indexedTypeName =  const indexedTypeName = this.current.typeChecker.typeToString(this.current.typeChecker.getTypeFromTypeNode(this.typeNode.type))
;
```

Move this into the catch block, we don't need the name of the type we are getting the keys from unless we throw.
In this case however, we want to include the name of the type we perform the keyof on.

Comment (Eywek) [reply src/metadataGeneration/typeResolver.ts]: Right sorry, should be fixed

**Comment by WoH** [comment]
LGTM. Thanks for your contribution :+1:

**Pull request merged**